packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -1448,6 +1475,16 @@ class _DateRangePickerDialogState extends State<DateRangePickerDialog> with Rest
     });
   }
 
+  bool _isDaySelectable(DateTime day) {
+    if (day.isBefore(widget.firstDate) || day.isAfter(widget.lastDate)) {
+      return false;
+    }
+    if (widget.selectableDayPredicate == null) {
+      return true;
+    }
+    return widget.selectableDayPredicate!(day, _selectedStart.value, _selectedEnd.value);
+  }
```

Comment:
@MitchellGoodwin Instead of making complex ifs, I've just refactored a bit to extract the selectable logic. At the end, it's much more understandable. wdyt ?

Reply:
You can see just above that the final code is way more understandable:

before:
```dart
if (_selectedStart.value != null &&
    (_selectedStart.value!.isBefore(widget.firstDate) || _selectedStart.value!.isAfter(widget.lastDate) ||
        widget.selectableDayPredicate != null && !widget.selectableDayPredicate!(_selectedStart.value!, _selectedStart.value, _selectedEnd.value))) {
  _selectedStart.value = null;
  // With no valid start date, having an end date makes no sense for the UI.
  _selectedEnd.value = null;
}
```
after:
```dart
if (selectedStart != null && !_isDaySelectable(selectedStart)) {
  _selectedStart.value = null;
  // With no valid start date, having an end date makes no sense for the UI.
  _selectedEnd.value = null;
}
```